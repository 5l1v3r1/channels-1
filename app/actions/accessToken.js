// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json.access_token)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST
  };
}

export function receiveAccessToken(token: string): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
    payload: token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE
  };
}
