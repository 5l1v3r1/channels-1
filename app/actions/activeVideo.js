// @flow
import actionTypes from '../constants/actionTypes';
import type { ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import RatingType from '../dataTypes/ratingType';

export function updateActiveVideo(video: any): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { id } = getState().activeVideo;
    if (id !== video.id) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_VIDEO,
        payload: video
      });
    }
  };
}

export function updateActiveVideoCounts(prevState: RatingType, rating: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    let likeCount = Number(getState().activeVideo.likeCount);
    let dislikeCount = Number(getState().activeVideo.dislikeCount);
    switch (rating) {
      case 'like':
        likeCount = prevState.like ? likeCount - 1 : likeCount + 1;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
        break;
      case 'dislike':
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount + 1;
        break;
      default:
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
    }
    dispatch({
      type: actionTypes.UPDATE_ACTIVE_VIDEO,
      payload: {
        likeCount: String(likeCount),
        dislikeCount: String(dislikeCount)
      }
    });
  };
}
