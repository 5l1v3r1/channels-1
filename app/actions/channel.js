// @flow
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

export function selectChannel(id: string): Action {
  return {
    type: actionTypes.SELECT_CHANNEL,
    payload: id
  };
}
