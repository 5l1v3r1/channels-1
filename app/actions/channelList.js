// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
