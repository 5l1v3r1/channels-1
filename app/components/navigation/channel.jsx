// @flow
import React from 'react';
import styles from './channel.scss';

const Channel = ({
  title,
  thumbnail,
  videoCount,
  subscriberCount,
  isActive,
  selectChannel
}: {
  title: string,
  thumbnail: string,
  videoCount: string,
  subscriberCount: string,
  isActive: boolean,
  selectChannel: (id: string) => void
}) => (
  <li className={isActive ? `${styles.channel} ${styles['channel--isActive']}` : styles.channel} onClick={selectChannel}>
    <img className={styles.thumbnail} src={thumbnail} />
    <div>
      <h3 className={styles.title}>{title}</h3>
      <span className={styles.subscribers}>{subscriberCount} subscribers</span>
      <span className={styles.videos}>{videoCount} videos</span>
    </div>
  </li>
);

export default Channel;
