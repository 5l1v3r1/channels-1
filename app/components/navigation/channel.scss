@import '../../app.global.scss';

.channel {
	display: flex;
	align-items: center;
	padding: 0.8rem 1rem;

	&--isActive {
		background-color: $gray;

		.subscribers,
		.videos {
			color: $white;
		}
	}
}

.thumbnail {
	align-self: center;
	width: 50px;
	margin-right: 0.8rem;
	border: 2px solid $white;
	border-radius: 50%;
	box-shadow: 0 1px 3px 1px rgba(46, 46, 46, 0.3);
}

.title {
	margin: 0;

	@include font-size(12);
}

.subscribers,
.videos {
	color: $gray;

	@include font-size(12);
}

.subscribers {
	margin-right: 0.8rem;
}
