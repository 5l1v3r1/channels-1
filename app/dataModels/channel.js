// @flow
import uuid from 'uuid';
import { stringOrEmpty, commaSeparateNumber } from '../utils/utils';

export default class Channel {
  title: string;
  thumbnail: string;
  videoCount: string;
  subscriberCount: string;
  isActive: boolean;
  id: string;

  constructor(
    title: string,
    thumbnail: string,
    videoCount: string,
    subscriberCount: string,
    isActive: boolean = false,
    id: string = uuid.v4()
  ) {
    this.title = title;
    this.thumbnail = thumbnail;
    this.videoCount = videoCount;
    this.subscriberCount = subscriberCount;
    this.isActive = isActive;
    this.id = id;
  }

  static from(json: any): Channel {
    const { snippet, statistics } = json;
    const title = stringOrEmpty(snippet.title);
    const thumbnail = stringOrEmpty(snippet.thumbnails.default.url);
    const videoCount = commaSeparateNumber(stringOrEmpty(statistics.videoCount));
    const subscriberCount = commaSeparateNumber(stringOrEmpty(statistics.subscriberCount));
    return new Channel(title, thumbnail, videoCount, subscriberCount);
  }
}
