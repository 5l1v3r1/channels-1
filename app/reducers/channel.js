// @flow
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';
import Channel from '../dataModels/channel';

export default function channel(state: Channel, { type, payload }: Action): Channel {
  switch (type) {
    case actionTypes.SELECT_CHANNEL:
      return new Channel(
        state.title,
        state.thumbnail,
        state.videoCount,
        state.subscriberCount,
        payload === state.id,
        state.id
      );
    default:
      return state;
  }
}
