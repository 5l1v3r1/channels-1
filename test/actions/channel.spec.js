import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as channelActions from '../../app/actions/channel';
import ChannelType from '../../app/dataTypes/channelType';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('channel actions', () => {
  it('creates TOGGLE_CHANNEL_LIST, UPDATE_ACTIVE_CHANNEL, UPDATE STATUS, and ENABLE_SEARCH_BAR when a channel is selected', () => {
    const store = mockStore({});
    store.dispatch(
      channelActions.selectChannel(
        new ChannelType(
          'DevTips',
          'https://yt3.ggpht.com/photo.jpg',
          '292',
          '232937',
          false,
          '8ef09d6c-1eda-4e1e-aeec-e3ddff47dd45'
        )
      )
    );
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should create an action to update the active channel', () => {
    const id = '8ef09d6c-1eda-4e1e-aeec-e3ddff47dd45';
    expect(channelActions.updateActiveChannel(id)).toMatchSnapshot();
  });
});
