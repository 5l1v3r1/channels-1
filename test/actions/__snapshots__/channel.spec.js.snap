// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`channel actions creates TOGGLE_CHANNEL_LIST, UPDATE_ACTIVE_CHANNEL, UPDATE STATUS, and ENABLE_SEARCH_BAR when a channel is selected 1`] = `
Array [
  Object {
    "type": "TOGGLE_CHANNEL_LIST",
  },
  Object {
    "payload": "8ef09d6c-1eda-4e1e-aeec-e3ddff47dd45",
    "type": "UPDATE_ACTIVE_CHANNEL",
  },
  Object {
    "payload": Object {
      "thumbnail": "https://yt3.ggpht.com/photo.jpg",
      "title": "DevTips",
    },
    "type": "UPDATE_STATUS",
  },
  Object {
    "payload": "DevTips",
    "type": "ENABLE_SEARCH_BAR",
  },
]
`;

exports[`channel actions should create an action to update the active channel 1`] = `
Object {
  "payload": "8ef09d6c-1eda-4e1e-aeec-e3ddff47dd45",
  "type": "UPDATE_ACTIVE_CHANNEL",
}
`;
